File: backend/app/modules/documents/service.py
"""Document ingestion: validate -> dedupe -> render/extract -> persist IR."""
from __future__ import annotations

from dataclasses import dataclass

from app.ai.ocr.base import OCREngine
from app.core.logging import get_logger
from app.db.models import Block, Document, Page
from app.db.repositories import BlockRepository, DocumentRepository, PageRepository
from app.db.session import UnitOfWork
from app.modules.documents.validation import validate_pdf
from app.modules.extraction.pipeline import extract_document
from app.schemas.common import DocumentKind
from app.schemas.ir import IRDocument
from app.storage.base import StorageBackend

log = get_logger(__name__)


@dataclass
class IngestResult:
    document: Document
    ir: IRDocument
    created: bool


class DocumentService:
    def __init__(self, session: UnitOfWork, storage: StorageBackend) -> None:
        self.session = session
        self.storage = storage
        self.documents = DocumentRepository(session)
        self.pages = PageRepository(session)
        self.blocks = BlockRepository(session)

    def ingest(
        self,
        *,
        organization_id: str,
        assessment_id: str,
        kind: DocumentKind,
        data: bytes,
        created_by: str | None = None,
        declared_mime: str | None = None,
        ocr_engine: OCREngine | None = None,
    ) -> IngestResult:
        validated = validate_pdf(data, declared_mime)

        log.info("ingest started", extra={"assessment_id": assessment_id, "kind": str(kind), "pages": validated.page_count})

        existing = self.documents.by_checksum(
            organization_id, assessment_id, str(kind), validated.checksum
        )
        if existing is not None:
            log.info("ingest skipped (duplicate)", extra={"assessment_id": assessment_id, "kind": str(kind), "document_id": existing.id})
            # Idempotent: re-uploading an identical file must not re-run OCR.
            return IngestResult(document=existing, ir=self.load_ir(existing), created=False)

        document = Document(
            organization_id=organization_id,
            created_by=created_by,
            assessment_id=assessment_id,
            kind=str(kind),
            storage_uri="",
            page_count=validated.page_count,
            mime=validated.mime,
            checksum=validated.checksum,
        )
        self.documents.add(document)
        meta = {
            "organization_id": organization_id,
            "assessment_id": assessment_id,
            "document_id": document.id,
            "kind": str(kind),
        }
        document.storage_uri = self.storage.put(
            f"{organization_id}/{assessment_id}/{document.id}/source.pdf", data, meta
        )

        output = extract_document(
            data,
            document_id=document.id,
            kind=str(kind),
            ocr_engine=ocr_engine,
            handwriting=kind == DocumentKind.ANSWER_SHEET,
        )

        for artifact in output.artifacts:
            uri = self.storage.put(
                f"{organization_id}/{assessment_id}/{document.id}/pages/{artifact.page_number}.png",
                artifact.image_bytes,
                {**meta, "page_number": artifact.page_number},
            )
            ir_page = output.ir.page(artifact.page_number)
            assert ir_page is not None
            ir_page.rendered_image_uri = uri

        document.ir_uri = self.storage.put(
            f"{organization_id}/{assessment_id}/{document.id}/ir.json",
            output.ir.model_dump_json(indent=2).encode(),
            meta,
        )
        document.markdown_uri = self.storage.put(
            f"{organization_id}/{assessment_id}/{document.id}/document.md",
            output.markdown.encode(),
            meta,
        )
        classifications = {page.classification for page in output.ir.pages}
        document.classification = (
            next(iter(classifications)) if len(classifications) == 1 else "mixed"
        )

        self._persist_ir(organization_id, created_by, document, output.ir)
        self.session.flush()
        log.info("ingest complete", extra={"assessment_id": assessment_id, "kind": str(kind), "document_id": document.id})
        return IngestResult(document=document, ir=output.ir, created=True)

    def _persist_ir(
        self, organization_id: str, created_by: str | None, document: Document, ir: IRDocument
    ) -> None:
        for ir_page in ir.pages:
            page = Page(
                organization_id=organization_id,
                created_by=created_by,
                document_id=document.id,
                page_number=ir_page.page_number,
                width=ir_page.width,
                height=ir_page.height,
                dpi=ir_page.dpi,
                classification=str(ir_page.classification),
                extraction_method=str(ir_page.extraction_method),
                rendered_image_uri=ir_page.rendered_image_uri,
            )
            self.pages.add(page)
            for ir_block in ir_page.blocks:
                self.blocks.add(
                    Block(
                        organization_id=organization_id,
                        created_by=created_by,
                        page_id=page.id,
                        block_key=ir_block.block_id,
                        text=ir_block.text,
                        bbox=ir_block.bbox.as_list(),
                        confidence=ir_block.confidence,
                        block_type=str(ir_block.block_type),
                        reading_order=ir_block.reading_order,
                        low_confidence=ir_block.low_confidence,
                    )
                )

    def load_ir(self, document: Document) -> IRDocument:
        if not document.ir_uri:
            return IRDocument(document_id=document.id, kind=document.kind, page_count=0, pages=[])
        raw = self.storage.get(document.ir_uri, organization_id=document.organization_id)
        return IRDocument.model_validate_json(raw.decode())

    def page_images(self, organization_id: str, document: Document) -> dict[int, bytes]:
        images: dict[int, bytes] = {}
        for page in self.pages.for_document(organization_id, document.id):
            if page.rendered_image_uri:
                images[page.page_number] = self.storage.get(
                    page.rendered_image_uri, organization_id=organization_id
                )
        return images

File: apps/web/lib/server/modules/documents/service.ts
/**
 * Document ingestion: validate -> dedupe -> render/extract -> persist IR.
 * Port of backend/app/modules/documents/service.py.
 */
import type { UnitOfWork } from "@/lib/server/db/session";
import { newOrgOwned } from "@/lib/server/db/base";
import { DocumentRepository, PageRepository, BlockRepository } from "@/lib/server/db/repositories";
import type { Document, Page, Block } from "@/lib/server/db/models";
import { GridFSStorage } from "@/lib/server/storage/gridfs";
import { validatePdf } from "./validation";
import { extractDocument } from "@/lib/server/modules/extraction/pipeline";
import { bboxToList } from "@/lib/server/modules/extraction/geometry";
import type { IRDocument } from "@/lib/server/modules/extraction/types";

export interface IngestResult {
  document: Document;
  ir: IRDocument;
  created: boolean;
}

export class DocumentService {
  private documents: DocumentRepository;
  private pages: PageRepository;
  private blocks: BlockRepository;

  constructor(private session: UnitOfWork, private storage: GridFSStorage) {
    this.documents = new DocumentRepository(session);
    this.pages = new PageRepository(session);
    this.blocks = new BlockRepository(session);
  }

  async ingest(opts: {
    organizationId: string;
    assessmentId: string;
    kind: "question_paper" | "answer_sheet";
    data: Buffer;
    createdBy?: string | null;
  }): Promise<IngestResult> {
    const { organizationId, assessmentId, kind, data, createdBy = null } = opts;
    const validated = await validatePdf(data);

    const existing = await this.documents.byChecksum(organizationId, assessmentId, kind, validated.checksum);
    if (existing) {
      // Idempotent: re-uploading an identical file must not re-run OCR.
      return { document: existing, ir: await this.loadIr(existing), created: false };
    }

    // `add()` tracks a shallow *copy* of the entity, not the object passed in — every
    // later mutation must go through the returned reference or it never reaches the
    // copy that flush() actually writes (this previously left storageUri/irUri/
    // markdownUri/classification stuck at their initial values forever, since all
    // four are only known after this point).
    const document: Document = this.documents.add({
      ...newOrgOwned(organizationId, createdBy),
      assessmentId,
      kind,
      storageUri: "",
      pageCount: validated.pageCount,
      mime: validated.mime,
      checksum: validated.checksum,
      classification: null,
      markdownUri: null,
      irUri: null,
    });

    const meta = { organizationId, assessmentId, documentId: document.id, kind };
    document.storageUri = await this.storage.put(
      `${organizationId}/${assessmentId}/${document.id}/source.pdf`,
      data,
      meta
    );

    const output = await extractDocument(data, document.id, kind, {
      handwriting: kind === "answer_sheet",
    });

    for (const artifact of output.artifacts) {
      const uri = await this.storage.put(
        `${organizationId}/${assessmentId}/${document.id}/pages/${artifact.pageNumber}.png`,
        artifact.imageBytes,
        { ...meta, pageNumber: artifact.pageNumber }
      );
      const irPage = output.ir.pages.find((p) => p.pageNumber === artifact.pageNumber);
      if (irPage) irPage.renderedImageUri = uri;
    }

    document.irUri = await this.storage.put(
      `${organizationId}/${assessmentId}/${document.id}/ir.json`,
      Buffer.from(JSON.stringify(output.ir, null, 2)),
      meta
    );
    document.markdownUri = await this.storage.put(
      `${organizationId}/${assessmentId}/${document.id}/document.md`,
      Buffer.from(output.markdown),
      meta
    );

    const classifications = new Set(output.ir.pages.map((p) => p.classification));
    document.classification = classifications.size === 1 ? [...classifications][0] : "mixed";

    await this.persistIr(organizationId, createdBy, document, output.ir);
    await this.session.flush();
    return { document, ir: output.ir, created: true };
  }

  private async persistIr(
    organizationId: string,
    createdBy: string | null,
    document: Document,
    ir: IRDocument
  ): Promise<void> {
    for (const irPage of ir.pages) {
      const page: Page = {
        ...newOrgOwned(organizationId, createdBy),
        documentId: document.id,
        pageNumber: irPage.pageNumber,
        width: irPage.width,
        height: irPage.height,
        dpi: irPage.dpi,
        classification: irPage.classification,
        extractionMethod: irPage.extractionMethod,
        renderedImageUri: irPage.renderedImageUri,
      };
      this.pages.add(page);
      for (const irBlock of irPage.blocks) {
        const block: Block = {
          ...newOrgOwned(organizationId, createdBy),
          pageId: page.id,
          blockKey: irBlock.blockId,
          text: irBlock.text,
          bbox: bboxToList(irBlock.bbox),
          confidence: irBlock.confidence,
          blockType: irBlock.blockType,
          readingOrder: irBlock.readingOrder,
          lowConfidence: irBlock.lowConfidence,
          script: irBlock.script,
          scriptScore: irBlock.scriptScore,
        };
        this.blocks.add(block);
      }
    }
  }

  async loadIr(document: Document): Promise<IRDocument> {
    if (!document.irUri) {
      return { documentId: document.id, kind: document.kind, pageCount: 0, pages: [] };
    }
    const raw = await this.storage.get(document.irUri, document.organizationId);
    return JSON.parse(raw.toString("utf-8")) as IRDocument;
  }

  async pageImages(organizationId: string, document: Document): Promise<Map<number, Buffer>> {
    const images = new Map<number, Buffer>();
    const pages = await this.pages.forDocument(organizationId, document.id);
    for (const page of pages) {
      if (page.renderedImageUri) {
        images.set(page.pageNumber, await this.storage.get(page.renderedImageUri, organizationId));
      }
    }
    return images;
  }
}
